Add /checkfree and /labfree (and other neat things)

nightmarishblue opened the pull request:
Added some functions, fixed some old things, and refactored (sorry)

nightmarishblue committed:
Set the groundwork for editing and testing using calendarbot's code. No real changes but finished getting .env to work with python for ease of command registration.

nightmarishblue committed:
Updated to Discord.JS v14. Will split the file probably.

nightmarishblue committed:
Split the JavaScript into multiple files. Also plan on making a full automator for registering the commands the bot needs.

nightmarishblue committed:
forgot to commit for like a month.  implemented error embeds and got requests for location events working.

nightmarishblue committed:
Forgot about this, roomcheck commands working now. Will get an automated command register script up.

nightmarishblue committed:
Finished the command management scripts.

nightmarishblue committed:
...forgot the README.

nightmarishblue committed:
Merge d local branch.

nightmarishblue committed:
The README is now warm and friendly to help you out.

nightmarishblue committed:
Fixed error with the daily messages.

nightmarishblue committed:
Fixed my muscle memory and turned console.err to console.error

nightmarishblue committed:
Merged my local branch with some fixes to the morning update stuff.

nightmarishblue committed:
Reminder that Node 17.5 is needed for fetch on the reg scripts.

DistroByte reviewed (changes requested):
Breaking from existing codebase style

I'm not sure we need to store all the IDs of the rooms, these are as likely to change as the actual timetable data.

Does the Discord version need to be updated? If not, why change it. Discord v13 is using the correct gateway, and if discord change that it'll be after I leave. Shouldn't be needed.

The register and delete bot command files should be automatic and shouldn't require input from the user.

I would advise renaming the command to `checkroom` and if possible return a timetable to show available hours in that room.

Comments go a long way also to improving readability - guilty on that front but I had no intention of anyone reading or modifying this code,

nightmarishblue committed:
Fixed issues with the morning update, I hope. Also, support for getting
tomorrow's more easily.

nightmarishblue committed:
Registration scripts require less user input now.

nightmarishblue committed:
Fixed issue where time would not be reset due to require()'s behaviour.
Chose not to switch to readFileSync.

nightmarishblue committed:
Squashed the last significant bugs, renamed the command and added
support for time ranges.

nightmarishblue commented:
Code has been commented and formatted nicely, I hope. 

Changed the command as specified to `checkrooms` and added support for custom time ranges in the underlying function. 

The registration scripts have been altered to need no input beyond execution.

As for obtaining room IDs, the room codes can't be easily queried in batches, or if they are I haven't found a way to make the opentimetable API do it. Either the site's keyword search can be performed for each one, or the API can be queried for campus location info, which has to be repeated for some 15 pages and scraped.

The smart thing would be to periodically periodically update the cache of room codes and their identities. I personally highly doubt existing timetable data will change for the rooms, but if you care enough I can add a scheduled task to generate the dictionary every once in a while.

Also, I fixed the channel sending feature, and added support for finding tomorrow's timetable info instead, although I don't know if you or anyone else would actually prefer that.

DistroByte commented:
Cheers for the work, I'll update the bot later this evening for you!
